lib/import_export/importer.rb

```diff
@@ -0,0 +1,160 @@
+
+module ImportExport
+  class Importer < ImportScripts::Base
+
+    def initialize(data)
+      @users = data[:users]
+      @groups = data[:groups]
+      @categories = data[:categories]
+      @topics = data[:topics]
+
+      # To support legacy `category_export` script
+      if data[:category].present?
+        @categories = [] if @categories.blank?
+        @categories << data[:category]
+      end
+    end
+
+    def perform
+      RateLimiter.disable
+
+      import_users
+      import_groups
+      import_categories
+      import_topics
+
+      self
+    ensure
+      RateLimiter.enable
+    end
+
+    def import_users
+      return if @users.blank?
+
+      @users.each do |u|
+        import_id = "#{u[:id]}#{import_source}"
+        existing = User.with_email(u[:email]).first
+
+        if existing
+          if existing.custom_fields["import_id"] != import_id
+            existing.custom_fields["import_id"] = import_id
+            existing.save!
+          end
+        else
+          u = create_user(u, import_id) # see ImportScripts::Base
+        end
+      end
+
+      self
+    end
+
+    def import_groups
+      return if @groups.blank?
+
+      @groups.each do |group_data|
+        g = group_data.dup
+        user_ids = g.delete(:user_ids)
+        external_id = g.delete(:id)
+        new_group = Group.find_by_name(g[:name]) || Group.create!(g)
+        user_ids.each do |external_user_id|
+          new_group.add(User.find(new_user_id(external_user_id))) rescue ActiveRecord::RecordNotUnique
+        end
+      end
+
+      self
+    end
+
+    def import_categories
+      return if @categories.blank?
+
+      import_ids = @categories.collect { |c| "#{c[:id]}#{import_source}" }
+      existing_categories = CategoryCustomField.where("name = 'import_id' AND value IN (?)", import_ids).select(:category_id, :value).to_a
+      existing_category_ids = existing_categories.pluck(:value)
+
+      @categories.reject! { |c| existing_category_ids.include? c[:id].to_s }
+      @categories.sort_by! { |c| c[:parent_category_id].presence || 0 }
+
+      @categories.each do |cat_attrs|
+        id = cat_attrs.delete(:id)
+        permissions = cat_attrs.delete(:permissions_params)
+
+        category = Category.new(cat_attrs)
+        category.parent_category_id = new_category_id(cat_attrs[:parent_category_id]) if cat_attrs[:parent_category_id].present?
+        category.user_id = new_user_id(cat_attrs[:user_id])
+        import_id = "#{id}#{import_source}"
+        category.custom_fields["import_id"] = import_id
+        category.permissions = permissions.present? ? permissions : { "everyone" => CategoryGroup.permission_types[:full] }
+        category.save!
+        existing_categories << { category_id: category.id, value: import_id }
+
+        if cat_attrs[:description].present?
+          post = category.topic.ordered_posts.first
+          post.raw = cat_attrs[:description]
+          post.save!
+          post.rebake!
```

Comment:
Do we need this? My understanding is that posts are automatically rebaked on save.

Reply:
I guess we are not doing rebake like `after_save: rebake!` or in any other way.

Reply:
Ooo my bad. Automatic rebake only happens when using `PostCreator`